File: gfg-weekly-coding-contest-74/BST Maximum Difference.java


//User function Template for Java

class Solution
{
    // Version 1: Find target number and get the minimum path sum from the target node
    // The difference in this problem is sum - path_sum rather than abs(sum - path_sum)
    // Therefore, we can just focus on minimum path sum.
    // TC: O(h) = O(n), SC: O(n)
    public static int maxDifferenceBST(Node root,int target)
    {
        //Please code here
        int sum = 0;
        boolean found = false;
        Node current = root;
        while(current != null) {
            if (current.data == target) {
                found = true;
                break;
            }
            sum += current.data;
            if (current.data > target) {
                current = current.left;
            } else {
                current = current.right;
            }
        }
        
        if (!found) return -1;
        int[] res = getValue(current);
        res[0] -= current.data;
        res[1] -= current.data;
        
        int ans = Integer.MIN_VALUE;
        if (res[0] != Integer.MAX_VALUE) {
            ans = Math.max(ans, sum - res[0]);
        }
        
        if (res[1] != 0) {
            ans = Math.max(ans, sum - res[1]);
        }
        
        return ans;
    }
    
    public static int[] getValue(Node node) {
        if (node == null) return new int[] {Integer.MAX_VALUE, 0};
        int[] res = new int[] {Integer.MAX_VALUE, 0};
        int[] left = getValue(node.left);
        int[] right = getValue(node.right);
            
        res[0] = Math.min(res[0], Math.min(left[0], right[0]));
        res[1] = Math.max(res[1], Math.max(left[1], right[1]));
        
        if (res[0] == Integer.MAX_VALUE) res[0] = 0;
        res[0] += node.data;
        res[1] += node.data;
        
        return res;
    }
}

File: gfg-weekly-coding-contest-74/BST Maximum Difference.py
#User function Template for python3

'''
class Node:
    def __init__(self,val):
        self.data=val
        self.left=None
        self.right=None
'''

class Solution:
    
    # Version 1: Find target number and get the minimum path sum from the target node
    # The difference in this problem is sum - path_sum rather than abs(sum - path_sum)
    # Therefore, we can just focus on minimum path sum.
    # TC: O(h) = O(n), SC: O(n)
    def maxDifferenceBST(self,root, target):
        #code here
        def dfs(node):
            if not node:
                return float('inf')
            res = min(dfs(node.left), dfs(node.right))
            if res == float('inf'):
                res = 0
            return res + node.data
        
        current = root
        total = 0
        while current:
            if current.data == target:
                break
            total += current.data
            if current.data > target:
                current = current.left
            else:
                current = current.right
        
        if not current:
            return -1
        return total - dfs(current) + current.data
